src/Parser/Template.cs

```diff
@@ -54,6 +57,95 @@ public Template(string template, bool isPositional, List<Literal> literals, List
       Holes = holes.ToArray();
     }
 
+    public string Render(object[] parameters)
+    {
+        var sb = new StringBuilder();
```

Comment:
I guess `Value.Length + 64 * Holes.Length` is a decent guess. Probably too large in many cases, but too large is better than too small and log messages are transient anyway (they are not meant to be promoted to GC gen 2).